### ISSUE TITLE
[BUG] - cardano-cli query ledger-state and protocol-state dump binary data

### ISSUE THREAD

**Issue by mkoura** [IntersectMBO/cardano-node#2461]
*Internal*

**Area**
*Other*   Any other topic (Delegation, Ranking, ...).

**Summary**
The `cardano-cli query ledger-state` and `cardano-cli query protocol-state` dump binary data when `--out-file` is specified. When `--out-file` is not specified and output is printed to stdout, the data is JSON-like as expected.

**Steps to reproduce**
1. run `cardano-cli query protocol-state --mary-era --testnet-magic 42 --cardano-mode --out-file protocol_state.log` and inspect the "protocol_state.log" file.

**Expected behavior**
Text JSON-like data instead of binary data.

**System info (please complete the following information):**
- OS: Fedora 33
- cardano-node 1.25.0 - linux-x86_64 - ghc-8.10
git rev 99e678c008405426f4af774f16f4f132539622f5

**Comment by kevinhammond** [comment]
I agree this is irritating, but the CLI code is a bit tricky here, and the design/implementation should be looked at as a whole to see what simplifications could be made.

A workaround is to redirect the standard output to a file.  Then `--out-file `is not needed.  

`cardano-cli query protocol-state --mary-era --testnet-magic 42 --cardano-mode > protocol_state.log
`
Longer term, it might be good to add a new flag that was specific to file output:

`--binary-format
`

**Comment by JaredCorduan** [comment]
The workaround no longer works for me on my laptop, since my OS kills the process due to memory consumption. it would be really great to have the JSON back.

**Comment by regel** [comment]
If we use `--out-file` flag and it writes binary then what are the tools, or command lines that can be used to decode the file? Could we produce the file dump on a node and then decode it offline on another computer?

**Comment by regel** [comment]
On testnet using `cardano-cli` version `1.29.1`, my Docker container gets killed if I limit memory to 1 GB and try to dump the ledger-state as a JSON file. It passes if I increase the memory limit to 2 GB.
Using `--out-file` to save the file in binary format is fine IMHO, but it means we need to have a new off-chain tool to decode this binary data. @JaredCorduan I started looking at the Cardano CLI source code and see where this type of feature can be plugged. I will update if I get something running.

**Comment by erikd** [comment]
If we still want to write the whole of the ledger state we can still use `Aeson` but be smarter about it. Rather than convert the whole ledger state struct to JSON and then write that out, an approach that will work when ledger state is really huge is to 
manually pull apart the ledger state and covert smaller chunks to JSON and then write them out individually, making sure the same JSON is written.

**Comment by JaredCorduan** [comment]
@shawnim what you've done is translate the CBOR _directly_ to JSON. which, as you can tell, is not nearly as human readable as the JSON produced by the CLI. The CLI converts the CBOR into a Haskell in-memory representation, which it then converts into JSON.

For example, a stake credential in CBOR might look like `[0, BYTES]`, but is rendered something like `{ "type": "keyhash", "value": HEX }` by the CLI. The `0` indicates a key hash (as opposed to `1` which indicates a script hash).

We have a CBOR specification for all the data that is transmitted on the wire (see [here](https://github.com/input-output-hk/cardano-ledger/tree/2048cd4679ef3155d88a445e0f818abad09a0827/eras/alonzo/test-suite/cddl-files) for Alonzo), but this does not include the entire ledger state.

**Comment by Jimbo4350** [comment]
Closing this. If this is still relevant please reopen.

**Comment by mkoura** [comment]
Still relevant.
